### DISCUSSION TITLE
packagename-stubs ignored?

### DISCUSSION THREAD

**Discussion by ptosco** [microsoft/pyright#5224] [Q&A] [answered]
Good morning, I have found this useful discussion: #3455 which pointed me to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/import-resolution.md).
I am particularly interested in point (3):

> Try to resolve using stubs or inlined types found within installed packages. Pyright uses the configured Python environment to determine whether a package has been installed. For more details about how to configure your Python environment for Pyright, see below. If a Python environment is configured, Pyright looks in the lib/site-packages, Lib/site-packages, or python*/site-packages subdirectory. If no site-packages directory can be found, Pyright attempts to run the configured Python interpreter and ask it for its search paths. If no Python environment is configured, Pyright will use the default Python interpreter by invoking python.

> For a given package, try to resolve first using a stub package. Stub packages, as defined in [PEP 561](https://www.python.org/dev/peps/pep-0561/#type-checker-module-resolution-order), are named the same as the original package but with “-stubs” appended.
Try to resolve using an inline stub, a “.pyi” file that ships within the package.

However, the above does not seem to work as described in VSCode, unless I set an explicit `python.analysis.stubPath` in `settings.json`.
My VSCode instance is configured to use `$CONDA_PREFIX/bin/python` as interpreter. Python packages are correctly resolved.

### The `-stubs` directory approach
I have created `$CONDA_PREFIX/lib/python3.8/site-packages/rdkit-stubs` directory:
```
$ rsync -rlt ~/typings/rdkit/ $CONDA_PREFIX/lib/python3.8/site-packages/rdkit-stubs
$ tree -L 1 $CONDA_PREFIX/lib/python3.8/site-packages/rdkit-stubs
$CONDA_PREFIX/lib/python3.8/site-packages/rdkit-stubs
├── Avalon
├── Chem
├── DataManip
├── DataStructs
├── Dbase
├── DistanceGeometry
├── ForceField
├── Geometry
├── __init__.pyi
├── ML
├── Numerics
├── rdBase
├── RDConfig
├── RDLogger
├── RDPaths
├── RDRandom
├── SimDivFilters
├── sping
├── TestRunner
├── utils
└── VLib
```
With this configuration, inlined types do not show up.
![image](https://github.com/microsoft/pyright/assets/5244385/cc4da7aa-826b-4e86-8a7e-9c8795650a89)

 ### The inline stub approach
I have merged the `$CONDA_PREFIX/lib/python3.8/site-packages/rdkit-stubs` into `$CONDA_PREFIX/lib/python3.8/site-packages/rdkit`, such that `.pyi` files are present next to `.py` files:
```
$ rsync -rlt ~/typings/rdkit/ $CONDA_PREFIX/lib/python3.8/site-packages/rdkit
$ tree -L 1 $CONDA_PREFIX/lib/python3.8/site-packages/rdkit
$CONDA_PREFIX/lib/python3.8/site-packages/rdkit
├── Avalon
├── Chem
├── cmake_install.cmake
├── CTestTestfile.cmake
├── DataManip
├── DataStructs
├── Dbase
├── DistanceGeometry
├── ForceField
├── Geometry
├── __init__.py
├── __init__.pyi
├── Makefile
├── ML
├── Numerics
├── __pycache__
├── rdBase
├── rdBase.so
├── RDConfig
├── RDConfig.py
├── RDLogger
├── RDLogger.py
├── RDPaths
├── RDPaths.py
├── RDRandom
├── RDRandom.py
├── SimDivFilters
├── six.py
├── sping
├── TestRunner
├── TestRunner.py
├── utils
└── VLib
```
Also with this configuration, inlined types do not show up.
![image](https://github.com/microsoft/pyright/assets/5244385/cc4da7aa-826b-4e86-8a7e-9c8795650a89)

### The separate `typings` directory approach with explicit `python.analysis.stubPath`

I created a `typings` directory containing the stub tree under `rdkit`:
```
$ rsync -rlt ~/typings $CONDA_PREFIX/lib/python3.8/site-packages
$ tree -L 1 $CONDA_PREFIX/lib/python3.8/site-packages/typings/rdkit
$CONDA_PREFIX/lib/python3.8/site-packages/typings/rdkit
├── Avalon
├── Chem
├── DataManip
├── DataStructs
├── Dbase
├── DistanceGeometry
├── ForceField
├── Geometry
├── __init__.pyi
├── ML
├── Numerics
├── rdBase
├── RDConfig
├── RDLogger
├── RDPaths
├── RDRandom
├── SimDivFilters
├── sping
├── TestRunner
├── utils
└── VLib
```
Then I added to `settings.json`:
```
{
    "python.analysis.stubPath": "$CONDA_PREFIX/lib/python3.8/site-packages/typings",
}
```
This yields the expected inline help:
![image](https://github.com/microsoft/pyright/assets/5244385/1435f136-702b-40b0-a57d-f6881a93d9bb)

## What I would like
Not having to set an explicit `python.analysis.stubPath` in `settings.json`.
Ideally I would like the `-stubs` directory approach to work as described in the documentation, such that I could distribute `rdkit-stubs` along with `rdkit` and users would not need to tweak their VSCode `settings.json`.

## Can someone help?
Many thanks in advance!

**Comment by erictraut** [comment, marked as answer]
Both of the approaches that you describe above in your "what does not work" section should work fine. Many other package authors use both of these approaches, so I'm sure they work fine in pyright. 

Some things for you to check:
1. Are you sure that you have the correct Python environment selected in VS Code? Make sure you've selected the environment where you've installed the stubs.
2. Try restarting the language server after you install the stubs in your environment. It's possible that the file system watcher that notifies pyright of changes to site-packages didn't work correctly, so pyright wasn't aware of the fact that you made a change. By restarting the language server, it will pick up the changes. The easiest way to restart the language server is by quitting VS Code and launching it again. Making a change to any of the pylance or pyright settings will also restart the language server.

Once you've installed your `rdkit-stubs` in the `site-packages` directory, you should be able to import from any of the `rdkit` submodules, and pyright will pick up the type information from the stubs. You can verify this by right clicking on an imported symbol and choosing "Go To Declaration". That should take you to the stub file that declares the symbol.

If you ultimately decide to ship the stubs inline with `rdkit` (as opposed to publishing a separate `rdkit-stubs` library), I recommend adding a `py.typed` marker file to your package. This will tell type checkers that your package supplies types. By default, pyright doesn't require that you include a `py.typed` marker in this scenario, but mypy does.

**Comment by ptosco** [comment]
@erictraut I figured out what the problem was: the `rdkit` project directory I am working on also contains an `rdkit` directory, but not a `rdkit-stubs` directory. Since in that case the project directory path has priority over `$CONDA_PREFIX/lib/python3.8/site-packages` for Python imports and it does **not** contain `rdkit-stubs`, inline help does not appear. Adding `rdkit-stubs` also in the project directory gives me inline help also when I am working in the `rdkit` project directory.

Thank you very much for your help - even though eventually I solved my own problem, as often happens your input helped me!

**Comment by ptosco** [comment]
@erictraut I am still missing something about return types.
If I look at `Chem.MolFromSmiles`, the return type is `Mol`:
![image](https://github.com/microsoft/pyright/assets/5244385/348e9254-bcac-42e6-ae08-562e91781a29)

However, VSCode thinks the return type is `Any`:
![image](https://github.com/microsoft/pyright/assets/5244385/a9c3c6eb-445b-48d0-bc9b-8df99bbd9003)

What am I missing?
I have tried both the `rdkit-stubs` approach and the `rdkit` approach with `__init__.pyi` files close to `__init__.py` files - neither works.
Adding `py.typed` makes no difference.

Thanks in advance for any help you may provide.

**Reply by erictraut** [reply]
@ptosco, I don't think I have enough information to help you diagnose the issue shown in the partial screen shots above. I'd need more context — and perhaps repro steps.

My best guess is that you've declared the type of the variable `mol` to be `Any`. Assigning a value to a variable declared as `Any` will not narrow the variable's type on assignment. If you change `mold = Chem.MolFromSmiles('CCC')` to `my_var = Chem.MolFromSmiles('CCC')` (where `my_var` is a local variable that has no type declaration), does it act as you expect?

**Comment by ptosco** [comment]
I have not declared `mol` at all in my test Python script.
This is the signature of `ChemMolFromSmiles` in my `.ipy` file:
```
def MolFromSmiles(SMILES: object, params: SmilesParserParams) -> Mol:
```
and in my Python script I only have:
```
from rdkit import Chem

mol = Chem.MolFromSmiles('CCC')
```
which yields to `Any` for `mol`:
![image](https://github.com/microsoft/pyright/assets/5244385/f043d951-d30e-4327-ab4c-63e4cc26fd5b)

However, if I try the same experiment with a Python function from the `subprocess` module, I get the expected type, rather than `Any`:
```
import subprocess

res = subprocess.run(["true"])
```
![image](https://github.com/microsoft/pyright/assets/5244385/854577ce-a394-42c8-8371-d51cb735132b)

The two expressions
```
mold = Chem.MolFromSmiles('CCC')
```
and
```
my_var = Chem.MolFromSmiles('CCC')
```
only differ for the variable name, so I am not sure why they should yield different outcomes (and they don't).

Thanks again for helping me to troubleshoot this.

**Reply by erictraut** [reply]
Once again, it's very difficult to diagnose issues like this with such piecemeal information. A self-contained, minimal code sample would be appreciated. If this code is checked in to a public repo (e.g. on an in-progress branch), that would also help.

One thing that I'll note is that the signature for `MolFromSmiles` that you've posted above specifies that two arguments are required, but you're passing it only one when you call it. That should generate a type violation error. I suspect that you have type checking disabled. If so, you should enable it because pyright will help guide you with error messages.

It's also unclear to me from the information you've provided whether `Chem` is the name of a class (as the capitalized symbol would imply) or a module (which would normally have a lowercase name). I suspect it's a module and that you're using non-standard naming conventions here because `MolFromSmiles` is a function that is named using class naming conventions.

**Comment by ptosco** [comment]
@erictraut I see your point.
I was able to reproduce the behavior in a freshly created conda environment following this workflow:
```
$ conda create -c conda-forge -n rdkit-pyright rdkit
$ conda activate rdkit-pyright
$ pip install pybind11-stubgen
$ curl -L -O https://raw.githubusercontent.com/ptosco/rdkit/gen_rdkit_stubs/Scripts/gen_rdkit_stubs.py
$ chmod +x gen_rdkit_stubs.py
$ ./gen_rdkit_stubs.py
[ ... you may ignore the console logging ... ]
```
Then point VSCode to the `rdkit-pyright` conda environment and trigger restart the Python Language Server by adding/removing an entry such as
```
    "python.defaultInterpreterPath": "/home/build/a3/envs/rdkit-pyright/bin/python",
```
In your `settings.json`.
You should then see the following:

```
from rdkit import Cherm

mol = Chem.MolFromSmiles('CCC')
```
<img width="418" alt="image" src="https://github.com/microsoft/pyright/assets/5244385/fbda9600-c7ba-48dd-a5e4-7e70a3453773">

and

<img width="194" alt="image" src="https://github.com/microsoft/pyright/assets/5244385/89717a51-3931-4f29-9be6-1388f1ab489d">

`Chem` is a module and `MolFromSmiles()` is a function.
RDKit is a C++ library, and Python biindings mostly follow C++ camel case naming conventions rather than Python conventions.

I really appreciate your help and responsiveness. I hope we can get to the bottom of this. Many thanks in advance again!

**Reply by erictraut** [reply]
Thanks for providing the repro steps, but I'm not able to get the `gen_rdkit_stubs.py` to work. It crashes before it outputs any stub files. I've spent about a half hour trying to debug it, but I can't afford to spend any more time on it. If you're able to provide alternate repro steps or make the script more robust so it runs without crashing, I'd be happy to take another look.

**Reply by ptosco** [reply]
@erictraut Sorry to hear that.
I have run the script successfully on two different machines (Linux and WSL). Do you get a stack trace? That might help me figure out.

**Reply by erictraut** [reply]
Perhaps you could upload the generated type stub file that defines the symbol `MolFromSmiles`? That way, we don't need to debug the stub generation script.

Based on your screen shot above, it looks like `MolFromSmiles` is defined to return an `Any` type. If you right click on `MolFromSmiles` and choose "Go To Declaration", it should take you to your generated stub file, and you can confirm what return type annotation is associated with it.

I'm also still confused by the fact that you are passing a single argument (`'CCC'`) to the function `Chem.MolFromSmiles` but the hover text indicates that the function has two parameters (`SMILES` and `params`). If you enable type checking errors in pyright, you should see a red squiggle and an error message indicating that you have not passed enough arguments for the function.

**Comment by ptosco** [comment]
Done, a tarfile is here:
[rdkit_pyi.tar.gz](https://raw.githubusercontent.com/ptosco/rdkit/gen_rdkit_stubs/Scripts/rdkit_pyi.tar.gz)
Expand it as follows:
```
$ cd $CONDA_PREFIX/lib/python3.11/site-packages
$ tar xzf rdkit_pyi.tar.gz
```
> Based on your screen shot above, it looks like `MolFromSmiles` is defined to return an `Any` type. If you right click on `MolFromSmiles` and choose "Go To Declaration", it should take you to your generated stub file, and you can confirm what return type annotation is associated with it.

This is the declaration that I am taken to when right clicking on `MolToSmiles`:
```
def MolFromSmiles(SMILES: object, params: SmilesParserParams) -> Mol:
```

> I'm also still confused by the fact that you are passing a single argument (`'CCC'`) to the function `Chem.MolFromSmiles` but the hover text indicates that the function has two parameters (`SMILES` and `params`).

I think this might have to do with the trouble I am having. These Python wrappers are generated from C++ code with `boost:::python`. The underlying C++ functions have multiple overloads taking various parameters, and `boost::python` generates wrappers for each overload taking advantage of Python's duck typing. However, apparently `pybind11-stubgen` only generates a single signature for one of the overloads. I need to investigate how to generate better signature parsing the docstrings myself.

**Reply by erictraut** [reply]
Thanks, I'm able to repro the problem now.

I think you've already identified your problem. The stubs you've generated are incorrect. They don't accurately reflect the interface to the rdkit library. 

As I've said above, I highly encourage you to enable type checking in pyright / pylance by setting `python.analysis.typeCheckingMode` to "basic". It appears that you have it set to "off", which is suppressing errors that would otherwise help you diagnose your problem.

Here's what it looks like when I enable type checking.

![image](https://github.com/microsoft/pyright/assets/7040122/4ee2daf8-ead3-4976-8245-c759252988fa)

![image](https://github.com/microsoft/pyright/assets/7040122/f212d01c-efae-4419-a570-1fa9dd40ecdb)

The reason that the type of the call expression evaluates to `Unknown` (which is an "implied `Any`" type) is because you haven't supplied appropriate arguments to the target function. The return type often depends on the argument types (e.g. if the call is generic or overloaded), so if you don't supply an argument list that's compatible with the signature of the function, pyright evaluates the type of the call expression as `Unknown`.

**Reply by ptosco** [reply]
@erictraut That makes sense. I hoped `pybind11-stubs` would do a better job OOTB, but apparently it doesn’t.
Thanks for helping me diagnose the problem. I hope I won’t need to bother you again.

**Reply by swamidass** [reply]
@ptosco, would you be willing to share/publish the rdkit-stubs files you've pulled together?

**Reply by swamidass** [reply]
Oh, see that you have, and that these stubs aren't right.

**Reply by ptosco** [reply]
@swamidass I have them 95% ready - I hope I will be able to release them in the next 2-3 weeks.
https://github.com/rdkit/UGM_2023/blob/main/Presentations/Tosco_A_stub_a_day_keeps_the_docstrings_at_bay.pdf

**Reply by swamidass** [reply]
Well that is exciting. I'd suggest releasing now with partial stubs! That would be a big improvement. From there, you can iterate!

If it is helpful, I wrote some code that's able to extract input signatures (unfortunately not output) from the exceptions that RDKit functions throw:

```

from __future__ import annotations

import lark
import typing


class ArgSpec(typing.NamedTuple):
    var: str
    type_: str
    default: str | None


def find_type(x, type_):
    return list(
        str(t)
        for t in x.scan_values(pred=lambda t: hasattr(t, "type") and t.type == type_)
    )

def data_to_str(tree, data: str):
  var = list(tree.find_data(data))
  assert len(var) <= 1
  var = "".join(var[0].scan_values(lambda x: True)) if var else None
  return var

    
def tree_to_declarations(tree):
    funcs = []
    for func in tree.find_data("function"):
        name = find_type(func, "FUNCTION")[0]

        args = []

        for arg in func.find_data("arg"):
            type_ = data_to_str(arg, "typespec") 
            var = data_to_str(arg, "var")
            default = data_to_str(arg, "default")

            args.append(ArgSpec(var, type_, default))

        funcs.append((name, args))
    return funcs


        
def get_error_text(func) -> str | None:
    try:
        func(1,2,3,4,5,6,76,7)

    except Exception as e:
        msg =  e.args[0]
        if "did not match C++ signature" not in msg:
          raise e from None
    
        return msg
    
error_grammar = r"""
start: error 

error: anytext function_line+


anytext: line (NEWLINE line)*
function_line.2: NEWLINE ws* function ws*
line: /.+/ 
function: FUNCTION "(" args ")"
lrvalue: "{lvalue}"| "{rvalue}"
ws: WS
subtypespec: NAME (ws+ NAME)? STAR? (LT ws* subtypespec ws* (COMMA ws* subtypespec ws*)* GT) ?
typespec: subtypespec

args: ws? arg ws? (comma ws? arg)* ws? 
arg: typespec  (ws+ lrvalue)? (ws+ var)?  ws* ("=" default)?

comma: ","


STAR: "*"
COMMA: ","
LT: "<"
GT: ">"
default: LITERAL
NAME: TOKEN
var.2: NAME
FUNCTION: NAME

TOKEN: LETTER ALLOWED_NAME_CHAR*

LITERAL: /[a-z0-1{}\.\[\]_]+/i
LETTER: /[a-z]/i
ALLOWED_NAME_CHAR:  /[a-z_0-9:\._]/i

%import common.ESCAPED_STRING
%import common.NEWLINE
%import common.WS
%ignore WS
"""


_L: lark.Lark = lark.Lark(error_grammar)

def inspect_rdkit_callable(func):
    
    err = get_error_text(func)

    tree = _L.parse(err)
    return tree_to_declarations(tree)


```

See these tests:

```
def test_var_none():
    M = Chem.MolFromSmiles("CCC")
    sig = inspect_rdkit_callable(M.GetIntProp)

    assert sig == [
        (
            "GetIntProp",
            [
                ArgSpec(var=None, type_="RDKit::ROMol const*", default=None),
                ArgSpec(var=None, type_="char const*", default=None),
            ],
        )
    ]

    sig = inspect_rdkit_callable(M.SetIntProp)

    assert sig == [
        (
            "SetIntProp",
            [
                ArgSpec(var="self", type_="RDKit::ROMol", default=None),
                ArgSpec(var="key", type_="char const*", default=None),
                ArgSpec(var="val", type_="int", default=None),
                ArgSpec(var="computed", type_="bool", default="False"),
            ],
        )
    ]


def test_overloaded():
    sig = inspect_rdkit_callable(Chem.MolFromSmiles)

    assert sig == [
    (
        'MolFromSmiles',
        [
            ArgSpec(var='SMILES', type_='boost::python::api::object', default=None),
            ArgSpec(var='sanitize', type_='bool', default='True'),
            ArgSpec(var='replacements', type_='boost::python::dict', default='{}')
        ]
    ),
    (
        'MolFromSmiles',
        [
            ArgSpec(var='SMILES', type_='boost::python::api::object', default=None),
            ArgSpec(var='params', type_='RDKit::SmilesParserParams', default=None)
        ]
    )
]

```
